### DISCUSSION TITLE
Using a volume amplitude envelope

### DISCUSSION THREAD

**Discussion by audio2000** [FluidSynth/fluidsynth#1376] [Q&A] [answered]
I apologize as I know very little about this.
There are various applications & plugins (LV2 VST) within Linux that use Fluidsynth API to handle soundfonts.  I am a little confused at how someone would be able to access a volume/amplitude envelope for controlling parameters such as attack, sustain, decay and release.

`APPLY TO ANY SOUNDFONT`
I was wondering if there was a set of these parameters that would allow them to be applied to ANY soundfont loaded in a Fluidsynth based player?

I have seen these mentioned in soundfont spec but have not been able to get them to work
```
26 attackModEnv
27 holdModEnv
28 decayModEnv 
29 sustainModEnv
30 releaseModEnv
```
For example I tried an application called LMMS that contains an instrument called SF2 Player which I believe uses FS.  This application provides you with a graphical MIDI CC array of knobs 0-127
When I try these CCs above I do not notice anything differences in sound.

On FS wiki I saw these on the list, did not see the other two decay & sustain, but regardless these parameters show X icons meaning they are not supported.  I tried them in LMMS as well but same result
```
(072) Sound Controller 3 (default: Release Time)
(073) Sound Controller 4 (default: Attack Time)
```

So does this mean that these parameters in the way I am looking for are not supported?

`ADD TO EACH SOUNDFONT IN THE FORM OF A MODULATOR`
I have also seen mentioned that in order to achieve this you have to add something called "modulators" inside each soundfont, which honestly does not sound that great if you wanted these parameters to be included in many many soundfonts.

Do you basically have to assign a "modulator" for each of these 4 parameters which get added inside the soundfont and then the CC numbers mentioned get linked to them somehow in the soundfont which then let MIDI access them?
The only soundfont editor I know of in Linux is Polyphone, the manual shows a brief section on modulators but not exactly sure how to achieve this.

If anyone has any information they could share about this, I would appreciate it.
If not too complext to explain, maybe an example of how to set a modulator for "release" parameter to a specific CC, I guess CC30 as it is shown above, I guess in Polyphone if possible.

Again I apologize for this question, I am just confused.
No instrument players / instrument formats (SF2, SFZ, GIG, DS) in Linux seem to offer these envelopes in a simple knob form, all seem to be needed to be added per instrument.

Thank You

Actually, one more thing
I have also seen people mention items, not specifically these but ones that look like this syntax
```
fluid_gen_type
fluid_synth_set_gen()
GEN_MODENVRELEASE 
```
I am assuming all items like this are for if someone is actually coding a FS based player, am I correct?

**Comment by derselbst** [comment, marked as answer]
> So does this mean that these parameters in the way I am looking for are not supported?

A default behavior for soundcontroller 3 and 4 is not defined by the soundfont spec. Hence, you need to apply modulator as you already mentioned. And yes, these modulator have to be applied to any soundfont and every instrument in that soundfont where you want to use it. That's not our invention, that's the standard way of how Soundfont was designed.

Yet, because this approach sucks, fluidsynth provides an API that would greatly simplify things: `fluid_synth_add_default_mod()`. Whether or not this API is adopted by the programs you're using is beyond my knowledge.

> Do you basically have to assign a "modulator" for each of these 4 parameters which get added inside the soundfont and then the CC numbers mentioned get linked to them somehow in the soundfont which then let MIDI access them?

Yes. Polyphone is the right tool for that. Start with a modulator that manipulates a generator that's easily audible, like `51coarseTune`.

> No instrument players / instrument formats (SF2, SFZ, GIG, DS) in Linux seem to offer these envelopes in a simple knob form, all seem to be needed to be added per instrument.

That's not possible, because nobody know how _you_ personally want the knob to behave: Linear, convex, what's the minimum, maximum, etc. - There's no standard or convention which defines that (to my best knowledge).

> I have also seen people mention items, not specifically these but ones that look like this syntax

This API is used for manipulating destination generators directly via NRPNs. Using this API directly is not suitable for your use-case. Ofc, you're free to use NRPN MIDI events to manipulate generators. I cannot recommend this for you currently. You should first make some initial steps to setup modulators, understand how they work and influence the various destination generators, before you step over to NRPNs.

**Comment by audio2000** [comment]
@derselbst 
Wow, perfectly answered, thank you very much for taking the time to answer each question specifically like this.
You do not have to go any further than this, what you have provided is enough, but

Just a clarification / slight follow up question, I think it is because phrased it like this 
> No instrument players / instrument formats (SF2, SFZ, GIG, DS) in Linux

that lead to your response
> That's not possible, because nobody know how you personally want the knob to behave: Linear, convex, what's the minimum, maximum, etc. - There's no standard or convention which defines that (to my best knowledge).

I was referring to instrument players for the any of these formats SF2, SFZ, GIG, DS do not contain players that provide these types of envelope parameters in Linux.  In Windows there are many, but do not want to use WINE.
I do not know what approach they are taking but here are some examples of the types of players I was referring to.

SF2
https://plugins4free.com/plugin/1089/
https://plugins4free.com/plugin/626/

SFZ
https://plugins4free.com/plugin/1703/
https://www.kvraudio.com/product/zampler-by-beat

Thank you again for answer each one of my questions
